File: kumbaja/transcript.py
"""Transcript reading, writing, and management."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .types import TurnRecord


class TranscriptManager:
    """Manages the deliberation transcript as a JSONL file."""

    def __init__(self, path: str) -> None:
        self.path = Path(path)
        self._records: list[TurnRecord] = []
        self._written = 0

    @property
    def records(self) -> list[TurnRecord]:
        return self._records

    def load_existing(self) -> list[TurnRecord]:
        """Load an existing transcript file for resume."""
        if not self.path.exists():
            return []
        loaded: list[TurnRecord] = []
        with open(self.path) as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    record = TurnRecord.from_json(line)
                    loaded.append(record)
                except (json.JSONDecodeError, KeyError):
                    continue
        self._records = loaded
        self._written = len(loaded)
        return loaded

    def append(self, record: TurnRecord) -> None:
        """Append a single record and write to disk."""
        self._records.append(record)
        # Write all unwritten records
        with open(self.path, "a") as f:
            for i in range(self._written, len(self._records)):
                f.write(self._records[i].to_json() + "\n")
            self._written = len(self._records)

    def write_all(self) -> None:
        """Rewrite the entire transcript file from memory."""
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with open(self.path, "w") as f:
            for record in self._records:
                f.write(record.to_json() + "\n")
        self._written = len(self._records)

    def last_agent_turn_index(self, agent_id: str) -> int | None:
        """Return the most recent turn index for a given agent, or None."""
        for i in range(len(self._records) - 1, -1, -1):
            if self._records[i].agent_id == agent_id:
                return i
        return None

    def history_for_agent(
        self,
        _agent_id: str,
        window: int,
        topology: Any,
        num_agents: int,
        turn: int,
    ) -> list[dict[str, str]]:
        """Build the history envelope for the next agent turn."""
        from .types import Topology

        if topology == Topology.STAR:
            # Star: see last K messages from any agent
            return [
                {"agent_id": r.agent_id, "content": r.content}
                for r in self._records[-window:]
            ]
        elif topology == Topology.MESH:
            # Mesh: see last K messages from any agent (same as star for envelope)
            return [
                {"agent_id": r.agent_id, "content": r.content}
                for r in self._records[-window:]
            ]
        else:
            # Ring: see last K messages from the predecessor agent
            if turn == 0:
                predecessor_id = "orchestrator"
            else:
                predecessor_idx = (turn - 1) % num_agents
                # Find the agent config for this index
                agent_order = self._infer_agent_order(num_agents)
                predecessor_id = agent_order[predecessor_idx]

            history = []
            for msg in reversed(self._records):
                if msg.agent_id == predecessor_id:
                    history.append({"agent_id": msg.agent_id, "content": msg.content})
                if len(history) >= window:
                    break
            history.reverse()
            return history

    def _infer_agent_order(self, num_agents: int) -> list[str]:
        """Infer the agent order from the first N non-orchestrator turns."""
        seen: list[str] = []
        for r in self._records:
            if r.agent_id == "orchestrator":
                continue
            if r.agent_id not in seen:
                seen.append(r.agent_id)
            if len(seen) >= num_agents:
                break
        return seen

    def consecutive_consensus_count(self) -> int:
        """Count how many consecutive turns have consensus markers."""
        count = 0
        for r in reversed(self._records):
            if r.consensus:
                count += 1
            else:
                break
        return count

    def total_cost(self) -> float:
        return sum(r.cost for r in self._records if r.cost is not None)

    def total_tokens(self) -> int:
        total = 0
        for r in self._records:
            if r.tokens.total:
                total += r.tokens.total
        return total

File: transcript.go
package kumbaja

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// TranscriptManager manages the deliberation transcript as a JSONL file.
type TranscriptManager struct {
	path    string
	records []TurnRecord
	written int
}

// NewTranscriptManager creates a new TranscriptManager for the given file path.
func NewTranscriptManager(path string) *TranscriptManager {
	return &TranscriptManager{path: path}
}

// Records returns the in-memory slice of all loaded/appended turn records.
func (tm *TranscriptManager) Records() []TurnRecord {
	return tm.records
}

// LoadExisting loads an existing JSONL transcript file into memory.
// Returns all loaded records or an empty slice if the file does not exist.
func (tm *TranscriptManager) LoadExisting() ([]TurnRecord, error) {
	if _, err := os.Stat(tm.path); os.IsNotExist(err) {
		return nil, nil
	}

	f, err := os.Open(tm.path)
	if err != nil {
		return nil, fmt.Errorf("opening transcript file: %w", err)
	}
	defer f.Close()

	var loaded []TurnRecord
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		var r TurnRecord
		if err := json.Unmarshal([]byte(line), &r); err != nil {
			continue // skip corrupted lines silently, matching Python behavior
		}
		loaded = append(loaded, r)
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("reading transcript file: %w", err)
	}

	tm.records = loaded
	tm.written = len(loaded)
	return loaded, nil
}

// Append appends a single record and writes all unwritten records to disk.
func (tm *TranscriptManager) Append(record TurnRecord) error {
	tm.records = append(tm.records, record)

	// Ensure the parent directory exists.
	if err := os.MkdirAll(filepath.Dir(tm.path), 0o755); err != nil {
		return fmt.Errorf("creating transcript directory: %w", err)
	}

	f, err := os.OpenFile(tm.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("opening transcript file for append: %w", err)
	}
	defer f.Close()

	for i := tm.written; i < len(tm.records); i++ {
		data, err := json.Marshal(tm.records[i])
		if err != nil {
			return fmt.Errorf("marshaling turn record: %w", err)
		}
		if _, err := f.Write(append(data, '\n')); err != nil {
			return fmt.Errorf("writing turn record: %w", err)
		}
	}
	tm.written = len(tm.records)
	return nil
}

// WriteAll rewrites the entire transcript file from memory.
func (tm *TranscriptManager) WriteAll() error {
	if err := os.MkdirAll(filepath.Dir(tm.path), 0o755); err != nil {
		return fmt.Errorf("creating transcript directory: %w", err)
	}

	f, err := os.Create(tm.path)
	if err != nil {
		return fmt.Errorf("creating transcript file: %w", err)
	}
	defer f.Close()

	for _, r := range tm.records {
		data, err := json.Marshal(r)
		if err != nil {
			return fmt.Errorf("marshaling turn record: %w", err)
		}
		if _, err := f.Write(append(data, '\n')); err != nil {
			return fmt.Errorf("writing turn record: %w", err)
		}
	}
	tm.written = len(tm.records)
	return nil
}

// HistoryForAgent builds the history envelope for the next agent turn.
//
// For star and mesh topologies, returns the last K messages from any agent.
// For ring topology, returns the last K messages from the predecessor agent only.
// At turn 0, the predecessor is "orchestrator".
// Otherwise, the predecessor is determined by (turn-1) % numAgents index
// using the agent order inferred from the first non-orchestrator turns.
func (tm *TranscriptManager) HistoryForAgent(agentID string, window int, topology Topology, numAgents int, turn int) []map[string]string {
	switch topology {
	case TopologyStar, TopologyMesh:
		// Star and mesh: last K messages from any agent.
		start := len(tm.records) - window
		if start < 0 {
			start = 0
		}
		var history []map[string]string
		for _, r := range tm.records[start:] {
			history = append(history, map[string]string{
				"agent_id": r.AgentID,
				"content":  r.Content,
			})
		}
		return history

	default:
		// Ring: last K messages from the predecessor agent.
		var predecessorID string
		if turn == 0 {
			predecessorID = "orchestrator"
		} else {
			predecessorIdx := (turn - 1) % numAgents
			agentOrder := tm.inferAgentOrder(numAgents)
			if predecessorIdx < len(agentOrder) {
				predecessorID = agentOrder[predecessorIdx]
			} else {
				// Fallback: not enough agents inferred yet, return empty.
				return nil
			}
		}

		var history []map[string]string
		for i := len(tm.records) - 1; i >= 0; i-- {
			if tm.records[i].AgentID == predecessorID {
				history = append(history, map[string]string{
					"agent_id": tm.records[i].AgentID,
					"content":  tm.records[i].Content,
				})
			}
			if len(history) >= window {
				break
			}
		}
		// Reverse to maintain chronological order.
		for i, j := 0, len(history)-1; i < j; i, j = i+1, j-1 {
			history[i], history[j] = history[j], history[i]
		}
		return history
	}
}

// inferAgentOrder infers the agent participation order from the first N
// non-orchestrator turns in the transcript.
func (tm *TranscriptManager) inferAgentOrder(numAgents int) []string {
	var seen []string
	for _, r := range tm.records {
		if r.AgentID == "orchestrator" {
			continue
		}
		found := false
		for _, s := range seen {
			if s == r.AgentID {
				found = true
				break
			}
		}
		if !found {
			seen = append(seen, r.AgentID)
		}
		if len(seen) >= numAgents {
			break
		}
	}
	return seen
}

// ConsecutiveConsensusCount counts how many consecutive trailing records
// have consensus=true.
func (tm *TranscriptManager) ConsecutiveConsensusCount() int {
	count := 0
	for i := len(tm.records) - 1; i >= 0; i-- {
		if tm.records[i].Consensus {
			count++
		} else {
			break
		}
	}
	return count
}

// TotalCost returns the sum of all non-nil cost values across records.
func (tm *TranscriptManager) TotalCost() float64 {
	var total float64
	for _, r := range tm.records {
		if r.Cost != nil {
			total += *r.Cost
		}
	}
	return total
}

// TotalTokens returns the sum of all non-nil total token counts across records.
func (tm *TranscriptManager) TotalTokens() int {
	total := 0
	for _, r := range tm.records {
		total += intVal(r.Tokens.Total)
	}
	return total
}
